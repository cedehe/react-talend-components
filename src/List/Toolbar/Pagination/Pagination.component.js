import React, { PropTypes } from 'react';
import classNames from 'classnames';
import uuid from 'uuid';
import { Nav, NavDropdown, MenuItem, ButtonGroup, Button } from 'react-bootstrap';
import Icon from '../../../Icon';

import css from './Pagination.scss';
/**
 * @param {object} props react props
 * @example
 <Pagination {...props}></Pagination>
 */

function Pagination({ id, itemsLength, sizeOptions, pageSize, activePage, onChangePagination }) {
	if (!pageSize || sizeOptions.indexOf(pageSize) < 0) {
		pageSize = sizeOptions[0];
	}
	const total = Math.ceil(itemsLength / pageSize);

	const changeSize = (value) => {
		onChangePagination(0, value);
	};

	const NEXT = 'next';
	const PREV = 'prev';
	const FIRST = 'first';
	const LAST = 'last';

	const navTo = (type) => {
		let from = activePage;
		switch (type) {
		case PREV: {
			if (from === 0) {
				return;
			}
			from -= 1;
			break;
		}
		case NEXT: {
			if (from === total - 1) {
				return;
			}
			from += 1;
			break;
		}
		case FIRST: {
			if (from === 0) {
				return;
			}
			from = 0;
			break;
		}
		case LAST: {
			if (from === total - 1) {
				return;
			}
			from = total - 1;
			break;
		}
		default:
			return;
		}
		onChangePagination(from * pageSize, pageSize);
	};
	const paginationSizeId = id && `${id}-pagination-size`;
	return (
		<Nav>
			<label className="navbar-text" htmlFor={paginationSizeId}>Show:</label>
			<NavDropdown
				id={paginationSizeId || uuid.v4()}
				title={pageSize}
				onSelect={changeSize}
			>
				{sizeOptions.map((option, index) => (
					<MenuItem key={index} eventKey={option}>{option}</MenuItem>
				))}
			</NavDropdown>
			<ButtonGroup className="navbar-btn">
				<Button
					id={id && `${id}-nav-to-first`}
					bsStyle="link"
					onClick={() => navTo(FIRST)}
					disabled={activePage === 0}
				>
					<Icon name="fa fa-backward" />
				</Button>
				<Button
					id={id && `${id}-nav-to-prev`}
					bsClass={classNames('btn btn-link', css['tc-pagination-ctrl-prev'], 'tc-pagination-ctrl-prev')}
					onClick={() => navTo(PREV)} disabled={activePage === 0}
				>
					<Icon name="fa fa-play" />
				</Button>
				<span className="btn btn-link">{activePage + 1}/{total}</span>
				<Button
					id={id && `${id}-nav-to-next`}
					bsStyle="link"
					onClick={() => navTo(NEXT)} disabled={activePage === (total - 1)}
				>
					<Icon name="fa fa-play" />
				</Button>
				<Button
					id={id && `${id}-nav-to-last`}
					bsStyle="link"
					onClick={() => navTo(LAST)} disabled={activePage === (total - 1)}
				>
					<Icon name="fa fa-forward" />
				</Button>
			</ButtonGroup>
		</Nav>
	);
}

Pagination.propTypes = {
	id: PropTypes.string,
	activePage: PropTypes.number,
	itemsLength: PropTypes.number.isRequired,
	onChangePagination: PropTypes.func.isRequired,
	pageSize: PropTypes.number,
	sizeOptions: PropTypes.arrayOf(PropTypes.number),
};

Pagination.defaultProps = {
	activePage: 0,
	sizeOptions: [5, 10, 20, 50],
};

export default Pagination;
