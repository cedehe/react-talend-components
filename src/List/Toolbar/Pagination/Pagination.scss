.tc-pagination-ctrl-prev {
  transform: rotate(180deg);
}
