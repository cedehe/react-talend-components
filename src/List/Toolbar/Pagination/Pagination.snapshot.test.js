import React from 'react';
import renderer from 'react-test-renderer';

import Pagination from './Pagination.component';

jest.mock('react-dom');

describe('Pagination', () => {
	it('should render', () => {
		// given
		const props = {
			itemsLength: 17,
			onChangePagination: jest.fn(),
		};
		
		// when
		const wrapper = renderer.create(
			<Pagination {...props} />
		).toJSON();
		
		// then
		expect(wrapper).toMatchSnapshot();
	});
	
	it('should render id if provided', () => {
		// given
		const props = {
			id: 'toolbar',
			itemsLength: 17,
			onChangePagination: jest.fn(),
		};
		
		// when
		const wrapper = renderer.create(
			<Pagination {...props} />
		).toJSON();
		
		// then
		expect(wrapper).toMatchSnapshot();
	});
});
