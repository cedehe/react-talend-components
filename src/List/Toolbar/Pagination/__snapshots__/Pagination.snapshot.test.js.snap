exports[`Pagination should render 1`] = `
<ul
  className="nav"
  role={null}>
  <label
    active={undefined}
    activeHref={undefined}
    activeKey={undefined}
    className="navbar-text"
    htmlFor={undefined}
    onSelect={null}>
    Show:
  </label>
  <li
    className="dropdown"
    style={undefined}>
    <a
      aria-expanded={false}
      aria-haspopup={true}
      className="dropdown-toggle"
      href=""
      id={42}
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button">
      5
       
      <span
        className="caret" />
    </a>
    <ul
      aria-labelledby={42}
      className="dropdown-menu"
      role="menu">
      <li
        className=""
        role="presentation"
        style={undefined}>
        <a
          href=""
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1">
          5
        </a>
      </li>
      <li
        className=""
        role="presentation"
        style={undefined}>
        <a
          href=""
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1">
          10
        </a>
      </li>
      <li
        className=""
        role="presentation"
        style={undefined}>
        <a
          href=""
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1">
          20
        </a>
      </li>
      <li
        className=""
        role="presentation"
        style={undefined}>
        <a
          href=""
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1">
          50
        </a>
      </li>
    </ul>
  </li>
  <div
    active={undefined}
    activeHref={undefined}
    activeKey={undefined}
    className="navbar-btn btn-group"
    onSelect={null}>
    <button
      className="btn btn-link"
      disabled={true}
      id={undefined}
      onClick={[Function]}
      type="button">
      <i
        aria-hidden="true"
        className="fa fa-backward"
        title={null} />
    </button>
    <button
      className="btn btn-link tc-pagination-ctrl-prev btn btn-link tc-pagination-ctrl-prev-default"
      disabled={true}
      id={undefined}
      onClick={[Function]}
      type="button">
      <i
        aria-hidden="true"
        className="fa fa-play"
        title={null} />
    </button>
    <span
      className="btn btn-link">
      1
      /
      4
    </span>
    <button
      className="btn btn-link"
      disabled={false}
      id={undefined}
      onClick={[Function]}
      type="button">
      <i
        aria-hidden="true"
        className="fa fa-play"
        title={null} />
    </button>
    <button
      className="btn btn-link"
      disabled={false}
      id={undefined}
      onClick={[Function]}
      type="button">
      <i
        aria-hidden="true"
        className="fa fa-forward"
        title={null} />
    </button>
  </div>
</ul>
`;

exports[`Pagination should render id if provided 1`] = `
<ul
  className="nav"
  role={null}>
  <label
    active={undefined}
    activeHref={undefined}
    activeKey={undefined}
    className="navbar-text"
    htmlFor="toolbar-pagination-size"
    onSelect={null}>
    Show:
  </label>
  <li
    className="dropdown"
    style={undefined}>
    <a
      aria-expanded={false}
      aria-haspopup={true}
      className="dropdown-toggle"
      href=""
      id="toolbar-pagination-size"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button">
      5
       
      <span
        className="caret" />
    </a>
    <ul
      aria-labelledby="toolbar-pagination-size"
      className="dropdown-menu"
      role="menu">
      <li
        className=""
        role="presentation"
        style={undefined}>
        <a
          href=""
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1">
          5
        </a>
      </li>
      <li
        className=""
        role="presentation"
        style={undefined}>
        <a
          href=""
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1">
          10
        </a>
      </li>
      <li
        className=""
        role="presentation"
        style={undefined}>
        <a
          href=""
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1">
          20
        </a>
      </li>
      <li
        className=""
        role="presentation"
        style={undefined}>
        <a
          href=""
          onClick={[Function]}
          onKeyDown={[Function]}
          role="menuitem"
          tabIndex="-1">
          50
        </a>
      </li>
    </ul>
  </li>
  <div
    active={undefined}
    activeHref={undefined}
    activeKey={undefined}
    className="navbar-btn btn-group"
    onSelect={null}>
    <button
      className="btn btn-link"
      disabled={true}
      id="toolbar-nav-to-first"
      onClick={[Function]}
      type="button">
      <i
        aria-hidden="true"
        className="fa fa-backward"
        title={null} />
    </button>
    <button
      className="btn btn-link tc-pagination-ctrl-prev btn btn-link tc-pagination-ctrl-prev-default"
      disabled={true}
      id="toolbar-nav-to-prev"
      onClick={[Function]}
      type="button">
      <i
        aria-hidden="true"
        className="fa fa-play"
        title={null} />
    </button>
    <span
      className="btn btn-link">
      1
      /
      4
    </span>
    <button
      className="btn btn-link"
      disabled={false}
      id="toolbar-nav-to-next"
      onClick={[Function]}
      type="button">
      <i
        aria-hidden="true"
        className="fa fa-play"
        title={null} />
    </button>
    <button
      className="btn btn-link"
      disabled={false}
      id="toolbar-nav-to-last"
      onClick={[Function]}
      type="button">
      <i
        aria-hidden="true"
        className="fa fa-forward"
        title={null} />
    </button>
  </div>
</ul>
`;
